### PR TITLE
Error when using dynamic attribute with a cyrillic code in a generic filter #3152

### PR DESCRIPTION
**Pull request by fractal3000**
…filter

### CODE REVIEW

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/JavaFieldNameSupportedSymbols.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import jakarta.validation.Payload;
+
+import java.lang.annotation.*;
+
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Constraint(validatedBy = JavaFieldNameSupportedSymbolsValidator.class)
+public @interface JavaFieldNameSupportedSymbols {
+    String message() default "The field can only contain latin letters or digits. First letter have to be a latin letter.";
```
Comment (Gavrilov-Ivan) [inline jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/JavaFieldNameSupportedSymbols.java:29]: Current default message is more strict than logic of the validator. Use the same text from the EN messages.properties.

**Review by Gavrilov-Ivan — CHANGES_REQUESTED** [review]

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/JavaFieldNameSupportedSymbols.java (hunk 1 above)
Comment (fractal3000) [reply jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/JavaFieldNameSupportedSymbols.java]: Fixed

**Review by Gavrilov-Ivan — APPROVED** [review]

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/JavaFieldNameSupportedSymbols.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import jakarta.validation.Constraint;
+import jakarta.validation.Payload;
+
+import java.lang.annotation.*;
+
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Constraint(validatedBy = JavaFieldNameSupportedSymbolsValidator.class)
+public @interface JavaFieldNameSupportedSymbols {
```
Comment (glebfox) [inline jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/JavaFieldNameSupportedSymbols.java:28]: You create an annotation for a specific attribute, so let's name it respectively. Now, it seems that I can apply it to any attribute and define any supported symbols by regexp, which is not true. Also, naming should be simple, like for other bean validation annotations, neither of them have `JavaField` in their name because it's defined show annotation is configured. 

After I wrote this, I realized that we have `jakarta.validation.constraints.Pattern`, so no need in custom annotation that uses regexp at all :-)

**Review by glebfox — CHANGES_REQUESTED** [review]

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/JavaFieldNameSupportedSymbols.java (hunk 2 above)
Comment (fractal3000) [reply jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/JavaFieldNameSupportedSymbols.java]: fixed

File: jmix-dynattr/dynattr/src/test/java/io/jmix/dynattr/model/CategoryAttributeTest.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ */
+
+package io.jmix.dynattr.model;
+
+import org.junit.jupiter.api.Test;
+
+import static io.jmix.dynattr.model.CategoryAttribute.CODE_FIELD_REGEXP;
+import static org.junit.jupiter.api.Assertions.*;
+
+class CategoryAttributeTest {
```
Comment (glebfox) [inline jmix-dynattr/dynattr/src/test/java/io/jmix/dynattr/model/CategoryAttributeTest.java:24]: We use Spock for new tests.

File: jmix-dynattr/dynattr/src/test/java/io/jmix/dynattr/model/CategoryAttributeTest.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        assertFalse(isValid(" abcd "));
+    }
+
+    @Test
+    void isValid_false_empty_string() {
+        assertFalse(isValid(""));
+    }
+
+    private static boolean isValid(String text){
+        return text.matches(CODE_FIELD_REGEXP);
```
Comment (glebfox) [inline jmix-dynattr/dynattr/src/test/java/io/jmix/dynattr/model/CategoryAttributeTest.java:98]: 1. We do not create API specifically for testing purposes. You don't use this constant in the entity. 
2. You don't test if entity validation works as expected, you test regexp. 

Get rid of constant and use `jakarta.validation.Validator` to validate an entity instance. See **Data Advanced** training course for the example of `Validator` usage.

File: jmix-translations/content/io/jmix/dynattr/messages_ru.properties [hunk 5]
```diff
@@ -15,6 +15,8 @@ io.jmix.dynattr.model/CategoryAttribute.category=Категория
 io.jmix.dynattr.model/CategoryAttribute.categoryEntityType=Тип категории
 io.jmix.dynattr.model/CategoryAttribute.name=Название
 io.jmix.dynattr.model/CategoryAttribute.code=Код
+io.jmix.dynattr.model/CategoryAttribute.code.validation.NotNull=Заполните код.
```
Comment (glebfox) [inline jmix-translations/content/io/jmix/dynattr/messages_ru.properties:18]: remove

File: jmix-dynattr/dynattr/src/main/resources/io/jmix/dynattr/messages.properties [hunk 6]
```diff
@@ -28,6 +28,8 @@ io.jmix.dynattr.model/CategoryAttribute.category=Category
 io.jmix.dynattr.model/CategoryAttribute.categoryEntityType=Category entity type
 io.jmix.dynattr.model/CategoryAttribute.name=Name
 io.jmix.dynattr.model/CategoryAttribute.code=Code
+io.jmix.dynattr.model/CategoryAttribute.code.validation.NotNull=Code required.
```
Comment (glebfox) [inline jmix-dynattr/dynattr/src/main/resources/io/jmix/dynattr/messages.properties:31]: remove

File: jmix-dynattr/dynattr/dynattr.gradle [hunk 7]
```diff
@@ -28,9 +28,12 @@ dependencies {
     testImplementation 'org.springframework:spring-test'
     testImplementation 'org.spockframework:spock-core'
     testImplementation 'org.spockframework:spock-spring'
+    testImplementation 'org.junit.jupiter:junit-jupiter-api'
```
Comment (glebfox) [inline jmix-dynattr/dynattr/dynattr.gradle:31]: We don't need this dependency if you'll use Spock

File: jmix-translations/content/io/jmix/dynattr/messages_ru.properties [hunk 8]
```diff
@@ -15,6 +15,8 @@ io.jmix.dynattr.model/CategoryAttribute.category=Категория
 io.jmix.dynattr.model/CategoryAttribute.categoryEntityType=Тип категории
```
Comment (glebfox) [inline jmix-translations/content/io/jmix/dynattr/messages_ru.properties:15]: You've missed the english translation in the `jmix-translations` module

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java [hunk 9]
```diff
@@ -186,7 +189,6 @@ public class CategoryAttribute implements Serializable {
     @Column(name = "ENUMERATION_LOCALES")
     protected String enumerationLocales;
 
-    @Column(name = "ATTRIBUTE_CONFIGURATION_JSON")
```
Comment (glebfox) [inline jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java:189]: why?

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java [hunk 10]
```diff
@@ -52,8 +52,10 @@
 public class CategoryAttribute implements Serializable {
     private static final long serialVersionUID = -6959392628534815752L;
 
+    static final String CODE_FIELD_REGEXP = "^[a-z]\\w*$";
```
Comment (glebfox) [inline jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java:55]: You don't need this constant

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java [hunk 11]
```diff
@@ -52,8 +52,10 @@
...
+
     @Id
-    @Column(name = "ID")
+    @Column(name = "ID", nullable = false)
```
Comment (glebfox) [inline jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java:58]: Get rid of unrelated changes

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java [hunk 12]
```diff
@@ -86,7 +88,7 @@ public class CategoryAttribute implements Serializable {
     private String deletedBy;
 
     @ManyToOne(optional = false, fetch = FetchType.LAZY)
-    @JoinColumn(name = "CATEGORY_ID")
+    @JoinColumn(name = "CATEGORY_ID", nullable = false)
```
Comment (glebfox) [inline jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java:91]: Get rid of unrelated changes

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java [hunk 13]
```diff
@@ -95,6 +97,8 @@ public class CategoryAttribute implements Serializable {
     @Column(name = "NAME", nullable = false)
     private String name;
 
+    @NotNull(message = "{msg://io.jmix.dynattr.model/CategoryAttribute.code.validation.NotNull}")
```
Comment (glebfox) [inline jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java:100]: 1. The default 'required' message is OK, no need in custom one. 
2. Get rid of unrelated change. If this field needs to be required at the DB level, then create a separate issue that will be implemented in the next minor release, since it requires a Liquibase migration script, which you've actually missed.

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java [hunk 14]
```diff
@@ -95,6 +97,8 @@ public class CategoryAttribute implements Serializable {
...
+    @Pattern(message = "{msg://io.jmix.dynattr.model/CategoryAttribute.code.validation.wrongCodeValueFormat}", regexp = "^[a-z]\\w*$")
```
Comment (glebfox) [inline jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java:101]: message already has `code` in this name, so no need to duplicate it. `msg://io.jmix.dynattr.model/CategoryAttribute.code.validation.wrongValueFormat` is OK.

**Review by glebfox — CHANGES_REQUESTED** [review]

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java (hunk 9 above)
Comment (fractal3000) [reply jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java:189]: It is unwilled mistake.

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java (hunk 11 above)
Comment (fractal3000) [reply jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java:58]: It is unwilled mistake. I haven't seen all changes just before commit.

File: jmix-dynattr/dynattr/src/test/java/io/jmix/dynattr/model/CategoryAttributeTest.java (hunk 4 above)
Comment (fractal3000) [reply jmix-dynattr/dynattr/src/test/java/io/jmix/dynattr/model/CategoryAttributeTest.java:98]: "Get rid of constant..." Why?

Comment (fractal3000) [reply jmix-dynattr/dynattr/src/test/java/io/jmix/dynattr/model/CategoryAttributeTest.java:98]: "You don't test if entity validation works as expected, you test regexp." I wanted to be shure that current behaviour and currenlty supporting cases wouldn't be broken with somebody's changes.

Comment (fractal3000) [reply jmix-dynattr/dynattr/src/test/java/io/jmix/dynattr/model/CategoryAttributeTest.java:98]: "We do not create API specifically for testing purposes." why? Sometimes it might be very usefull. There are special testing frameworks tha implements API specially for testing. Futhemore the code of the method is trivial.

File: jmix-dynattr/dynattr/dynattr.gradle (hunk 7 above)
Comment (fractal3000) [reply jmix-dynattr/dynattr/dynattr.gradle:31]: Is there any example of implementing unit test for entity with bean validation engine?

File: jmix-dynattr/dynattr/dynattr.gradle [hunk 15]
```diff
@@ -33,4 +33,5 @@ dependencies {
     testRuntimeOnly 'org.junit.vintage:junit-vintage-engine'
     testRuntimeOnly 'org.slf4j:slf4j-simple'
     testRuntimeOnly 'org.hsqldb:hsqldb'
+
```
Comment (glebfox) [inline jmix-dynattr/dynattr/dynattr.gradle:36]: Unrelated changes

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java [hunk 16]
```diff
@@ -117,7 +118,6 @@ public class CategoryAttribute implements Serializable {
             @AttributeOverride(name = "intEntityId", column = @Column(name = "DEFAULT_INT_ENTITY_VALUE")),
             @AttributeOverride(name = "longEntityId", column = @Column(name = "DEFAULT_LONG_ENTITY_VALUE"))
     })
-    @EmbeddedParameters(nullAllowed = true)
```
Comment (glebfox) [inline jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java:120]: Is it unwilled mistake? Please, review changes before commiting.

**Review by glebfox — CHANGES_REQUESTED** [review]

File: jmix-dynattr/dynattr/dynattr.gradle (hunk 15 above)
Comment (fractal3000) [reply jmix-dynattr/dynattr/dynattr.gradle]: Fixed

File: jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java (hunk 14 above)
Comment (fractal3000) [reply jmix-dynattr/dynattr/src/main/java/io/jmix/dynattr/model/CategoryAttribute.java]: Agree. Fixed

**Review by glebfox — APPROVED** [review]
